### PR TITLE
Struct parse and enhancement in tests

### PR DESCRIPTION
**Pull request by Jeconias**
### :sparkles: Enhancements

* Improved testing during struct parse

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,5 @@
 # go-my-ast-hurts
+
 A helper for dealing with Go AST
+
+**THIS IS A WORK IN PROGRESS**
```
Comment (felipemfp) [inline README.md]: image

File: definitions.go [hunk 2]
```diff
@@ -93,21 +95,39 @@ type MethodArgument struct {
...
 	Params []TagParam
 }
 
+func (t *Tag) AppendTagParam(tNew *TagParam) bool {
+	_, ok := t.TagParamByName(tNew.Name)
+	if ok {
+		return !ok
+	}
+	t.Params = append(t.Params, *tNew)
+	return ok
```
Comment (felipemfp) [inline definitions.go]: Isn't this going to be always false at this point? Is that the desired behavior?

Comment (Jeconias) [reply definitions.go]: False inform you that there was no problem in add new value, however, the return type should change to error. Thx 😁

File: experiment_test.go [hunk 3]
```diff
@@ -2,7 +2,6 @@ package myasthurts_test
 
 import (
 	"fmt"
-	"go/ast"
 	"go/parser"
 	"go/token"
```
Comment (edumarcal) [inline experiment_test.go]: Replace from `jamillosantos` to `lab259` in line 9.


*E.g.:* `myasthurts "github.com/lab259/go-my-ast-hurts"`

File: experiment_test.go [hunk 4]
```diff
@@ -14,6 +13,174 @@ import (
...
+		})
+
+		It("should parse struct fields", func() {
+			fset := token.NewFileSet()
+			f, err := parser.ParseFile(fset, "data/models2.sample", nil, parser.AllErrors)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(f).ToNot(BeNil())
+			Expect(f.Decls).ToNot(BeNil())
+
+			//ast.Print(fset, f)
```
Comment (edumarcal) [inline experiment_test.go]:
```suggestion
```

File: experiment_test.go [hunk 5]
```diff
@@ -14,6 +13,174 @@ import (
...
+			Expect(pkg.Structs[1].Fields).To(HaveLen(2))
+
+			Expect(pkg.Structs[0].Fields[0].Name).To(Equal("ID"))
+			Expect(pkg.Structs[0].Fields[1].Name).To(Equal("Name"))
+			Expect(pkg.Structs[0].Fields[2].Name).To(Equal("Email"))
+
+			Expect(pkg.Structs[0].Fields[0].Type).ToNot(Equal(BeNil()))
+			Expect(pkg.Structs[0].Fields[1].Type).ToNot(Equal(BeNil()))
+			Expect(pkg.Structs[0].Fields[2].Type).ToNot(Equal(BeNil()))
+
```
Comment (edumarcal) [inline experiment_test.go]:
```suggestion
```

File: experiment_test.go [hunk 6]
```diff
@@ -14,6 +13,174 @@ import (
...
+			Expect(pkg.Structs[0].Fields[0].Type.Name).To(Equal("int64"))
+			Expect(pkg.Structs[0].Fields[1].Type.Name).To(Equal("string"))
+			Expect(pkg.Structs[0].Fields[2].Type.Name).To(Equal("string"))
+
+			Expect(pkg.Structs[0].Fields[0].Type.Type).To(BeNil())
+			Expect(pkg.Structs[0].Fields[1].Type.Type).To(BeNil())
+			Expect(pkg.Structs[0].Fields[2].Type.Type).To(BeNil())
+
```
Comment (edumarcal) [inline experiment_test.go]:
```suggestion
```

File: experiment_test.go [hunk 7]
```diff
@@ -14,6 +13,174 @@ import (
...
+			Expect(pkg.Structs[1].Fields[0].Name).To(Equal("ID"))
+			Expect(pkg.Structs[1].Fields[1].Name).To(Equal("Address"))
+
+			Expect(pkg.Structs[1].Fields[0].Type).ToNot(Equal(BeNil()))
+			Expect(pkg.Structs[1].Fields[1].Type).ToNot(Equal(BeNil()))
+
```
Comment (edumarcal) [inline experiment_test.go]:
```suggestion
```

File: experiment_test.go [hunk 8]
```diff
@@ -14,6 +13,174 @@ import (
...
+			Expect(pkg.Structs[1].Fields[0].Type.Name).To(Equal("int64"))
+			Expect(pkg.Structs[1].Fields[1].Type.Name).To(Equal("string"))
+
+			Expect(pkg.Structs[1].Fields[0].Type.Type).To(BeNil())
+			Expect(pkg.Structs[1].Fields[1].Type.Type).To(BeNil())
+
```
Comment (edumarcal) [inline experiment_test.go]:
```suggestion
```

File: experiment_test.go [hunk 9]
```diff
@@ -14,6 +13,174 @@ import (
...
+			pkg, ok := env.PackageByName("models")
+			Expect(ok).To(BeTrue())
+
+			Expect(pkg).NotTo(BeNil())
+			Expect(pkg.Structs).To(HaveLen(1))
+			Expect(pkg.Structs[0].Name()).To(Equal("User"))
+			Expect(pkg.Structs[0].Fields).To(HaveLen(3))
+
+			Expect(pkg.Structs[0].Fields[0].Tag).NotTo(BeNil())
+			Expect(pkg.Structs[0].Fields[0].Tag.Raw).To(Equal("json:\"id\""))
```
Comment (edumarcal) [inline experiment_test.go]:
```suggestion
			Expect(pkg.Structs[0].Fields[0].Tag.Raw).To(Equal(`json:"id"`))
```

File: experiment_test.go [hunk 10]
```diff
@@ -14,6 +13,174 @@ import (
...
+			Expect(pkg.Structs[0].Fields[0].Tag.Params).To(HaveLen(1))
+			Expect(pkg.Structs[0].Fields[0].Tag.Params[0].Name).To(Equal("json"))
+			Expect(pkg.Structs[0].Fields[0].Tag.Params[0].Value).To(Equal("id"))
+			Expect(pkg.Structs[0].Fields[0].Tag.Params[0].Options).To(BeEmpty())
+
+			Expect(pkg.Structs[0].Fields[1].Tag.Raw).To(Equal("json:\"name,lastName\""))
```
Comment (edumarcal) [inline experiment_test.go]:
```suggestion
			Expect(pkg.Structs[0].Fields[1].Tag.Raw).To(Equal(`json:"name,lastName"`))
```

File: experiment_test.go [hunk 11]
```diff
@@ -14,6 +13,174 @@ import (
...
+			Expect(pkg.Structs[0].Fields[1].Tag.Params).To(HaveLen(1))
+			Expect(pkg.Structs[0].Fields[1].Tag.Params[0].Name).To(Equal("json"))
+			Expect(pkg.Structs[0].Fields[1].Tag.Params[0].Value).To(Equal("name"))
+			Expect(pkg.Structs[0].Fields[1].Tag.Params[0].Options).To(HaveLen(1))
+			Expect(pkg.Structs[0].Fields[1].Tag.Params[0].Options[0]).To(Equal("lastName"))
+
+			Expect(pkg.Structs[0].Fields[2].Tag.Raw).To(Equal("json:\"address\" bson:\"\""))
```
Comment (edumarcal) [inline experiment_test.go]:
```suggestion
			Expect(pkg.Structs[0].Fields[2].Tag.Raw).To(Equal(`json:"address" bson:""`))
```

File: experiment_test.go [hunk 12]
```diff
@@ -14,6 +13,174 @@ import (
...
+		})
+
+		It("should parse struct custom field", func() {
+			fset := token.NewFileSet()
+			f, err := parser.ParseFile(fset, "data/models4.sample", nil, parser.AllErrors)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(f).ToNot(BeNil())
+			Expect(f.Decls).ToNot(BeNil())
+
+			//ast.Print(fset, f)
```
Comment (edumarcal) [inline experiment_test.go]:
```suggestion
```

File: experiment_test.go [hunk 13]
```diff
@@ -14,6 +13,174 @@ import (
...
+		})
+
+		It("should parse struct comments", func() {
+			fset := token.NewFileSet()
+			f, err := parser.ParseFile(fset, "data/models5.sample", nil, parser.ParseComments)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(f).ToNot(BeNil())
+			Expect(f.Decls).ToNot(BeNil())
+
+			//ast.Print(fset, f)
```
Comment (edumarcal) [inline experiment_test.go]:
```suggestion
```

File: experiment_test.go [hunk 14]
```diff
@@ -145,6 +312,10 @@ var _ = Describe("My AST Hurts", func() {
 		// TODO
 	})
 
+	/*Context("Struct Tags", func () {
+		//
+	})*/
+
```
Comment (edumarcal) [inline experiment_test.go]:
```suggestion
```

File: experiment_test.go [hunk 15]
```diff
@@ -436,31 +607,81 @@ var _ = Describe("My AST Hurts", func() {
...
+		env := myasthurts.NewEnvironment()
+		myasthurts.Parse(env, f)
+
+		pkg, ok := env.PackageByName("models")
+		Expect(ok).To(BeTrue())
+
+		Expect(pkg.Structs).To(HaveLen(1))
+		Expect(pkg.Structs[0].Comment).To(HaveLen(2))
+
```
Comment (edumarcal) [inline experiment_test.go]: We should assert comments

File: experiment_test.go [hunk 16]
```diff
@@ -436,31 +607,81 @@ var _ = Describe("My AST Hurts", func() {
...
+		Expect(ok).To(BeTrue())
+
+		Expect(pkg.Structs).To(HaveLen(1))
+		Expect(pkg.Structs[0].Comment).To(BeEmpty())
+		Expect(pkg.Structs[0].Fields).To(HaveLen(2))
+		Expect(pkg.Structs[0].Fields[1].Comment).To(HaveLen(2))
+
+	})
+
+	/*It("should parse comments in struct", func() {
```
Comment (edumarcal) [inline experiment_test.go]: Why is this test commented?

File: definitions.go (hunk 2 above)
Comment (Jeconias) [reply definitions.go]: This has been resolved in branch **builtin-parse**. I believe than not is necessary return error, so, i changed func to resolve problem with return. (always returning false).

File: experiment_test.go [hunk 17]
```diff
@@ -14,6 +13,174 @@ import (
 
 var _ = Describe("My AST Hurts", func() {
 
+	FContext("should parse struct", func() {
```
[the remark this answers is not in the document]
Comment (Jeconias) [reply experiment_test.go]: Resolved in branch **builtin-parse**.

**Comment by Jeconias** [comment]
All suggestion of @edumarcal and @felipemfp has been resolved in branch **builtin-parse**.

**Pull request merged**